### PR TITLE
bpo-22367: Add open_file_descriptor parameter to fcntl.lockf()

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue22367

### CODE REVIEW

File: Modules/fcntlmodule.c [hunk 1]
```diff
@@ -362,12 +363,18 @@ starts.  `whence` is as with fileobj.seek(), specifically:
     0 - relative to the start of the file (SEEK_SET)
     1 - relative to the current buffer position (SEEK_CUR)
     2 - relative to the end of the file (SEEK_END)
+
+The default for `start` is 0, which means to start at the beginning of the file.
```
Comment (serhiy-storchaka) [inline Modules/fcntlmodule.c]: It is redundant since default values are shown in the signature.

File: Modules/fcntlmodule.c [hunk 2]
```diff
@@ -417,6 +424,16 @@ fcntl_lockf_impl(PyObject *module, int fd, int code, PyObject *lenobj,
...
+#ifdef F_OFD_SETLK
+                l.l_pid = 0;
+                ret = fcntl(fd, (code & LOCK_NB) ? F_OFD_SETLK : F_OFD_SETLKW, &l);
+#else
+                PyErr_SetString(PyExc_NotImplementedError,
+                                "lockf: open_file_descriptor is not supported on this platform");
+                return NULL;
+#endif
+            } else
             ret = fcntl(fd, (code & LOCK_NB) ? F_SETLK : F_SETLKW, &l);
```
Comment (serhiy-storchaka) [inline Modules/fcntlmodule.c]: Indentation and braces.

**Review by serhiy-storchaka — COMMENTED** [review]
In addition I would add a test based on the example https://www.gnu.org/software/libc/manual/html_node/Open-File-Description-Locks-Example.html

**Comment by corona10** [comment]
@serhiy-storchaka 
> In addition I would add a test based on the example https://www.gnu.org/software/libc/manual/html_node/Open-File-Description-Locks-Example.html

Thanks for the review.
I will add unit test based on the link.

**Comment by corona10** [comment]
@serhiy-storchaka @vstinner 
I 've added the unit test based on https://www.gnu.org/software/libc/manual/html_node/Open-File-Description-Locks-Example.html.
Please take a look :)

**Comment by corona10** [comment]
I left a question on https://bugs.python.org/issue22367 which related to the parameter name.

File: Doc/library/fcntl.rst [hunk 3]
```diff
@@ -155,6 +155,12 @@ The module defines the following functions:
    The default for *len* is 0 which means to lock to the end of the file.  The
    default for *whence* is also 0.
 
+   The default for *open_file_descriptor* is False, if *open_file_descriptor* is
+   set to True, open file description locks are used.
```
Comment (pitrou) [inline Doc/library/fcntl.rst]: This is cryptic. Few people know what an "open file description lock" is. Can you perhaps link to a manual page or some kind of doc on the topic?

Comment (pitrou) [reply Doc/library/fcntl.rst]: Also, if this is Linux-specific, then it should also be mentioned (e.g. "open file description locks are a Linux-specific feature").

Comment (corona10) [reply Doc/library/fcntl.rst]: @pitrou I agree

**Comment by corona10** [comment]
@pitrou @serhiy-storchaka @vstinner

Thanks for the review.
I 've updated the PR.

@vstinner please take a look at my suggestion on https://bugs.python.org/msg356329 which suggests changing the parameter name from `open_file_descriptor` to `open_file_description`.

**Comment by corona10** [comment]
@vstinner If you don't mind can you leave some feedback on this PR?

**Comment by corona10** [comment]
@vstinner Let's reopen this PR when we need this feature again ;)

**Pull request closed without merging**